smack-core/src/main/java/org/jivesoftware/smack/altConnections/HttpLookupMethod.java

```diff
@@ -0,0 +1,187 @@
+        /**
+         * Selects link relation attribute as "urn:xmpp:alt-connections:xbosh".
+         */
+        BOSH("urn:xmpp:alt-connections:xbosh"),
+        /**
+         * Selects link relation attribute as "urn:xmpp:alt-connections:websocket".
+         */
+        WEBSOCKET("urn:xmpp:alt-connections:websocket");
+        private final String attribute;
+        LinkRelation(String relAttribute) {
+            this.attribute = relAttribute;
+        }
+    }
+
+    /**
+     * Constructs a HTTP connection with the host specified by the DomainBareJid
+     * and retrieves XRD element in the form of an InputStream.
+     *
+     * @param xmppServiceAddress address of host
+     * @return InputStream containing XRD element
+     * @throws IOException exception due to input/output operations
+     */
+    public static InputStream getXrdStream(DomainBareJid xmppServiceAddress) throws IOException {
+        final String metadataUrl = "https://" + xmppServiceAddress.toString() + "/.well-known/host-meta";
+        final URL putUrl = new URL(metadataUrl);
+        final URLConnection urlConnection = putUrl.openConnection();
+        return  urlConnection.getInputStream();
+    }
+
+    /**
+     * Get remote endpoints for the provided LinkRelation from provided XmlPullParser.
+     *
+     * @param parser XmlPullParser that contains LinkRelations
+     * @param relation type of endpoints specified by the given LinkRelation
+     * @return list of endpoints
+     * @throws IOException exception due to input/output operations
+     * @throws XmlPullParserException exception encountered during XML parsing
+     * @throws URISyntaxException exception to indicate that a string could not be parsed as a URI reference
+     */
+    public static List<URI> parseXrdLinkReferencesFor(XmlPullParser parser, String relation) throws IOException, XmlPullParserException, URISyntaxException {
+        final String xrd_namespace = "http://docs.oasis-open.org/ns/xri/xrd-1.0";
+
+        List<URI> uriList = new ArrayList<>();
+        int initialDepth = parser.getDepth();
+
+        XmlPullParser.TagEvent tag = parser.nextTag();
+        loop: while (true) {
+            switch (tag) {
+            case START_ELEMENT:
+                String name = parser.getName();
+                String namespace = parser.getNamespace();
+                String rel = parser.getAttributeValue("rel");
+
+                if (!namespace.equals(xrd_namespace) || !name.equals("Link") || !rel.equals(relation)) {
+                    tag = parser.nextTag();
+                    continue loop;
+                }
+                String endpointUri = parser.getAttributeValue("href");
+                uriList.add(new URI(endpointUri));
+                tag = parser.nextTag();
+                break;
+            case END_ELEMENT:
+                if (parser.getDepth() == initialDepth) {
+                    break loop;
+                }
+                tag = parser.nextTag();
```

Comment:
That's an unusually parsing loop design. IIRC the call to nextTag() is once right at the start of the loop, and not twice, one time before the loop and in the end element case.